[mlir][ROCDL] Refactor wmma intrinsics to use attributes not operands where possible

Muzammiluddin-Syed-ECE opened the pull request:
The current implementation of the WMMA intrinsic ops as they are defined in the ROCDL tablegen is incorrect. They represent as operands what should be attributes such as `clamp`, `opsel`, `signA/signB`. This change performs a refactoring to bring it in line with what we expect.

Muzammiluddin-Syed-ECE committed:
[mlir][ROCDL] refactor wmma intrinsics to use attributes instead of operands where possible

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

Muzammiluddin-Syed-ECE committed:
removing unnecessary changes

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

kuhar commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -1311,11 +1300,33 @@ struct WMMAOpLowering : public ConvertOpToLLVMPattern<WMMAOp> {
+    // bitcast them back.
+    auto aType = dyn_cast<VectorType>(adaptor.getSourceA().getType());
+    auto bType = dyn_cast<VectorType>(adaptor.getSourceA().getType());
+    auto destCType = dyn_cast<VectorType>(adaptor.getDestC().getType());
```
either `cast` if you know it must be vector type or `dyn_cast` and check

Also, is there no `getSourceAType()` in the adapter?

kuhar commented on mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:
```diff
@@ -582,51 +582,148 @@ def ROCDL_smfmac_f32_32x32x32_fp8_fp8 : ROCDL_Mfma_IntrOp<"smfmac.f32.32x32x32.f
+             LLVM_ScalarOrVectorOf<CD>:$C)> {
+  let results = (outs LLVM_ScalarOrVectorOf<CD>:$res);
+  let assemblyFormat = [{
+    $A `,` $B `,` $C attr-dict `:` `(`type($A) `,` type($B) `,` type($C)`)` `->` type($res)
```
I think we can do `functional-type(operands, $res)`, no?

Muzammiluddin-Syed-ECE committed:
[mlir][ROCDL] Address review comments

- Use SmallVectorImpl for output parameters per LLVM coding standards
- Fix type casting bug (getSourceA -> getSourceB on line 1309)
- Use cast<> instead of dyn_cast<> for guaranteed vector types
- Simplify assembly format with functional-type shorthand

Addresses review comments from @kuhar on PR #167041

Muzammiluddin-Syed-ECE committed:
[mlir][ROCDL] Fix Python bindings test for WMMA refactoring

Update the Python test to match the new WMMA API where operands are
passed as separate positional arguments and attributes (like opsel) are
passed as keyword arguments with boolean values.

Changes:
- Changed from list-style arguments to positional arguments
- Changed opsel from MLIR Value operand to Python bool attribute
- Removed unnecessary false constant creation

This fixes the CI test failure where the Python bindings were expecting
the old API signature.

kuhar reviewed (approved):
LGTM but wait for @krzysz00 to review as well

krzysz00 commented on mlir/test/Conversion/AMDGPUToROCDL/wmma-gfx11.mlir:
```diff
@@ -13,23 +13,23 @@ func.func @wmma_to_rocdl(%arg0 : vector<16xf16>, %arg1 : vector<8xf32>, %arg2 :
   // CHECK-NEXT: llvm.bitcast %[[raw_bf16x8]] : vector<8xi16> to vector<8xbf16>
   amdgpu.wmma 16x16x16 %arg3 * %arg3 + %arg5 {subwordOffset = 0 : i32}: vector<16xbf16>, vector<16xbf16>, vector<8xbf16>
-  // CHECK: rocdl.wmma.i32.16x16x16.iu8{{.*}}: (i1, vector<4xi32>, i1, vector<4xi32>, vector<8xi32>, i1) -> vector<8xi32>
+  // CHECK: rocdl.wmma.i32.16x16x16.iu8{{.*}}: (vector<4xi32>, vector<4xi32>, vector<8xi32>) -> vector<8xi32>
```
... Hold on. This isn't preserving `clamp`. Or at least not testing for it.

krzysz00 reviewed (commented):
Let's ensure we have (working, precise) tests for all the things that just got made attributes with their non-default values, otherwise LGTM here

Muzammiluddin-Syed-ECE committed:
Addressing review comments

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

krzysz00 reviewed (approved):
LGTM now
